File: tests/tests_mashnetwork/binary_server.cpp
#include <mash-network/server.h>
#include <mash-network/server_listener.h>
#include <mash-utils/stringutils.h>
#include "binary_data.h"

using namespace Mash;
using namespace std;


class BinaryListener: public ServerListener
{
public:
    BinaryListener(int socket)
    : ServerListener(socket)
    {
        char buffer1[50];
        char buffer2[50];

        sprintf(buffer1, "BinaryListener #%d", socket);
        sprintf(buffer2, "logs/tests/BinaryListener_%d_$TIMESTAMP.log", socket);
        
        _outStream.open(buffer1, buffer2);
    }

    virtual ~BinaryListener()
    {
    }

    virtual tAction handleCommand(const std::string& strCommand,
                                  const ArgumentsList& arguments)
    {
        if (strCommand != "SEND_DATA")
        {
            if (!sendResponse("UNKNOWN_COMMAND", ArgumentsList()))
                return ACTION_CLOSE_CONNECTION;

            return ACTION_NONE;
        }

        if (!sendResponse("DATA", ArgumentsList(DATA_size)))
            return ACTION_CLOSE_CONNECTION;

        if (!sendData(DATA, DATA_size))
            return ACTION_CLOSE_CONNECTION;

        unsigned char* data = new unsigned char[DATA_size];

        if (!waitData(data, DATA_size))
            return ACTION_CLOSE_CONNECTION;

        for (unsigned int i = 0; i < DATA_size; ++i)
        {
            if (DATA[i] != data[i])
                return ACTION_CLOSE_CONNECTION;
        }

        if (!sendResponse("TEST_OK", ArgumentsList()))
            return ACTION_CLOSE_CONNECTION;

        return ACTION_NONE;
    }
};


ServerListener* createListener(int socket)
{
    return new BinaryListener(socket);
}



int main(int argc, char** argv)
{
    unsigned int nbMaxClients = 0;
    
    if (argc == 2)
        nbMaxClients = StringUtils::parseUnsignedInt(argv[1]);
    
    Server server(nbMaxClients);
    
    server.listen("127.0.0.1", 10000, createListener);
    
    return 0;
}

File: tests/tests_mashnetwork/binary_server.py
#! /usr/bin/env python

################################################################################
# The MASH Framework contains the source code of all the servers in the
# "computation farm" of the MASH project (http://www.mash-project.eu),
# developed at the Idiap Research Institute (http://www.idiap.ch).
#
# Copyright (c) 2016 Idiap Research Institute, http://www.idiap.ch/
# Written by Philip Abbet (philip.abbet@idiap.ch)
#
# This file is part of the MASH Framework.
#
# The MASH Framework is free software: you can redistribute it and/or modify
# it under the terms of either the GNU General Public License version 2 or
# the GNU General Public License version 3 as published by the Free
# Software Foundation, whichever suits the most your needs.
#
# The MASH Framework is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public Licenses
# along with the MASH Framework. If not, see <http://www.gnu.org/licenses/>.
################################################################################


import sys
import os


# Parameters handling
if (len(sys.argv) > 3) or ((len(sys.argv) == 2) and (sys.argv[1] == '--help')):
    print "Usage: %s [MODULES_PATH [NB_MAX_CLIENT]]" % sys.argv[0]
    sys.exit(-1)

if len(sys.argv) == 3:
    sys.path.append(sys.argv[1])
    nb_max_clients = int(sys.argv[2])
elif len(sys.argv) == 2:
    sys.path.append(sys.argv[1])
    nb_max_clients = 0
else:
    nb_max_clients = 0

from pymash import ServerListener, Server, Message

sys.path.append(os.path.dirname(sys.argv[0]))

import binary_data
    

class BinaryListener(ServerListener):

    def handleCommand(self, command):
        if command.name != "SEND_DATA":
            if not(self.sendResponse("UNKNOWN_COMMAND")):
                return ServerListener.ACTION_CLOSE_CONNECTION
        else:
            if not(self.sendResponse(Message("DATA", [binary_data.DATA_size]))):
                return ServerListener.ACTION_CLOSE_CONNECTION

            if not(self.sendData(''.join(map(lambda x: chr(x), binary_data.DATA)))):
                return ServerListener.ACTION_CLOSE_CONNECTION

            data = self.waitData(binary_data.DATA_size)

            if data is None:
                return ServerListener.ACTION_CLOSE_CONNECTION

            for i in range(0, binary_data.DATA_size):
                if binary_data.DATA[i] != ord(data[i]):
                    return ServerListener.ACTION_CLOSE_CONNECTION

            if not(self.sendResponse("TEST_OK")):
                return ServerListener.ACTION_CLOSE_CONNECTION

        return ServerListener.ACTION_NONE


server = Server(nb_max_clients)
server.run('127.0.0.1', 10000, BinaryListener)
